File: app/src/main/java/com/ruuvi/station/model/HumidityCalibration.java
package com.ruuvi.station.model;

import android.support.annotation.Nullable;

import com.ruuvi.station.database.tables.RuuviTagEntity;

import java.util.Date;
import java.util.HashMap;

public class HumidityCalibration {

    private static HashMap<String, HumidityCalibration> cache = new HashMap<>();

    public String mac;
    public double humidityOffset;
    public Date timestamp = new Date();

    public static HumidityCalibration calibrate(RuuviTagEntity tag) {
        HumidityCalibration prevCalibration = HumidityCalibration.get(tag);
        double prevCalibrationValue = 0f;
        if (prevCalibration != null) {
            prevCalibrationValue = prevCalibration.humidityOffset;
        }
        double calibration = 75f - (tag.getHumidity() != null ? tag.getHumidity() : 0.0 - prevCalibrationValue);
        HumidityCalibration newCalibration = new HumidityCalibration();
        newCalibration.humidityOffset = calibration;
        newCalibration.mac = tag.getId();
        cache.put(tag.getId(), newCalibration);
        return newCalibration;
    }

    public static void clear(RuuviTagEntity tag) {
        cache.remove(tag.getId());
    }

    public static RuuviTagEntity apply(RuuviTagEntity tag) {
        HumidityCalibration calibration = get(tag);
        if (calibration != null) {
            tag.setHumidity(tag.getHumidity() + calibration.humidityOffset);
        }
        return tag;
    }

    @Nullable
    public static HumidityCalibration get(RuuviTagEntity tag) {
        if (cache.containsKey(tag.getId())) {
            return cache.get(tag.getId());
        }
        return null;
    }
}

File: app/src/main/java/com/ruuvi/station/model/HumidityCalibration.kt
package com.ruuvi.station.model

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager

class HumidityCalibration {
    companion object {
        var cache = HashMap<String, Float>()
        fun calibrate(context: Context, tag: RuuviTag): RuuviTag {
            val pref: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
            val prevCalibration = this.get(context, tag)
            val calibration = tag.humidity.toFloat()+prevCalibration-75f
            pref.edit().putFloat("humidity_calibration:"+tag.id, calibration).apply()
            cache.put(tag.id, calibration)
            tag.humidity -= calibration
            return tag
        }
        fun clear(context: Context, tag: RuuviTag) {
            val pref: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
            pref.edit().putFloat("humidity_calibration:"+tag.id, 0f).apply()
            cache.remove(tag.id)
        }
        fun apply(context: Context, tag: RuuviTag): RuuviTag {
            val calibration = this.get(context, tag)
            tag.humidity -= calibration
            return tag
        }
        fun get(context: Context, tag: RuuviTag): Float {
            if (cache.contains(tag.id)) {
                return cache.get(tag.id)!!
            }
            val pref: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
            return pref.getFloat("humidity_calibration:"+tag.id, 0f)
        }
    }
}